### ISSUE TITLE
toolchain '@local_config_cc//:cc-compiler-x64_windows'

### ISSUE THREAD

**Issue by pemgithub** [google/gapid#2914]
I'm following the Windows build directions ( https://github.com/google/gapid/blob/master/BUILDING.md ), and I get an error.
```
ERROR: C:/users/pemgithub/_bazel_pemgithub/3mdmxsrc/external/io_bazel_rules_go/BUILD.bazel:62:1: Target '@io_bazel_rules_go//:go_context_data' depends on toolchain '@local_config_cc//:cc-compiler-x64_windows', which cannot be found: no such target '@local_config_cc//:cc-compiler-x64_windows': target 'cc-compiler-x64_windows' not declared in package '' defined by C:/users/pemgithub/_bazel_pemgithub/3mdmxsrc/external/local_config_cc/BUILD.bazel'
```
I think maybe the Windows build directions are missing a step on how to install toolchain '@local_config_cc//:cc-compiler-x64_windows'.  choco installed bazel version 0.27.1

**Comment by AWoloszyn** [comment]
I wonder if this is a bazel version issue. It is possible you can get around this by specifying a specific bazel version (0.26.0 and 0.27.0 both work I think).

If my memory serves me right:
`choco install bazel --version=0.26.1 --allow-downgrade`

**Comment by pemgithub** [comment]
(choco install bazel --version=0.26.1) fails to install.
(choco install bazel --version=0.27.0) gets the aforementioned :cc-compiler-x64_windows error

**Comment by pemgithub** [comment]
Could it be proxy related?  I'm new to bazel but...  I read to set HTTP_PROXY (and HTTPS_PROXY) environment variables, but repository.bzl -> github_http_args() was still failing, so I modified my local copy of repository.bzl -> github_http_args() to use a local copy of zip downloads.

The full error message is:
```
ERROR: C:/users/pemgithub/_bazel_pemgithub/3mdmxsrc/external/bazel_tools/tools/cpp/BUILD:58:1: Target '@bazel_tools//tools/cpp:malloc' depends on toolchain '@local_config_cc//:cc-compiler-x64_windows', which cannot be found: no such target '@local_config_cc//:cc-compiler-x64_windows': target 'cc-compiler-x64_windows' not declared in package '' defined by C:/users/pemgithub/_bazel_pemgithub/3mdmxsrc/external/local_config_cc/BUILD.bazel'
ERROR: D:/pemcode/amd/gapid/cmd/gapir/cc/BUILD.bazel:34:1: Target '//cmd/gapir/cc:gapir_unstripped' depends on toolchain '@local_config_cc//:cc-compiler-x64_windows', which cannot be found: no such target '@local_config_cc//:cc-compiler-x64_windows': target 'cc-compiler-x64_windows' not declared in package '' defined by C:/users/pemgithub/_bazel_pemgithub/3mdmxsrc/external/local_config_cc/BUILD.bazel'
ERROR: Analysis of target '//:pkg' failed; build aborted: Analysis failed
INFO: Elapsed time: 0.897s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded, 0 targets configured)
    currently loading: @io_bazel_rules_go//go/private ... (3 packages)
    Fetching @com_github_golang_protobuf; Restarting.
```

**Comment by pemgithub** [comment]
(choco install bazel --version=0.26.0) seems to be working better...  trying that

**Comment by pemgithub** [comment]
(bazel build pkg -s) gets farther with bazel 0.26.0...  But I still get errors...
```
ERROR: C:/users/pemgithub/_bazel_pemgithub/3mdmxsrc/external/gapic_third_party/BUILD.bazel:41:1: no such package '@org_lwjgl_core//jar': Failed to fetch Maven dependency: Could not transfer artifact org.lwjgl:lwjgl:jar:3.2.0 from/to org_lwjgl_core (https://repo1.maven.org/maven2/): connect timed out and referenced by '@gapic_third_party//:lwjgl'
ERROR: Analysis of target '//:pkg' failed; build aborted: no such package '@org_lwjgl_core//jar': Failed to fetch Maven dependency: Could not transfer artifact org.lwjgl:lwjgl:jar:3.2.0 from/to org_lwjgl_core (https://repo1.maven.org/maven2/): connect timed out
INFO: Elapsed time: 11.858s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded, 0 targets configured)
    Fetching @com_github_pkg_errors; Restarting. 10s
```

**Comment by pemgithub** [comment]
Btw I also get the following (unrelated to maven proxy bug).  Probably GAPID should update to support latest androidndk (version 20)?
```
WARNING: The major revision of the Android NDK referenced by android_ndk_repository rule 'androidndk' is 20. The major revisions supported by Bazel are [10, 11, 12, 13, 14, 15, 16, 17, 18]. Bazel will attempt to treat the NDK as if it was r18. This may cause compilation and linkage problems. Please download a supported NDK version.
```

**Comment by hevrard** [comment]
Make sure you installed Android SDK/NDK by yourself, as recommended in:
https://github.com/google/gapid/blob/master/BUILDING.md#install-android-sdk-and-ndk
and update the environment variables.

I recommend NDK version 18b, which is the one we currently use in our CI. You can see our Windows CI script here:
https://github.com/google/gapid/blob/master/kokoro/windows/build.bat
NOTE: it is tailored to the windows environment in our CI, in particular it does NOT use chocolatey, so I still recommend to refer to the Building documentation to build by hand, but this script may help you.

Can you double check which version of Bazel you are using now using `bazel --version`?

**Comment by null77** [comment]
Can someone from the gapid team take a crack at this? I think this is a legit break in the Windows build.

**Comment by pemgithub** [comment]
I think it's related to needing to tell bazel to use proxy settings (behind corporate firewall) for maven external dependencies.  Or I could just download the *.jar files manually, but idk how to tell bazel to do that (or how to concisely edit GAPID's bazel files to do that).

**Comment by pemgithub** [comment]
$ bazel version
Build label: 0.26.0
Build target: bazel-out/x64_windows-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Tue May 28 08:36:37 2019 (1559032597)
Build timestamp: 1559032597
Build timestamp as int: 1559032597

**Comment by pemgithub** [comment]
I extracted https://dl.google.com/android/repository/android-ndk-r18b-windows-x86_64.zip to D:\Programs\android-ndk-r18b, then (set ANDROID_NDK_HOME=D:\Programs\android-ndk-r18b) before (bazel build pkg).  I get the same error.  I think it's related to maven being blocked by corporate firewall, so I think I need to tell bazel to use corporate proxy for maven, or to use local copy of jar files (that I can download from maven)
```
ERROR: C:/users/pemgithub/_bazel_pemgithub/3mdmxsrc/external/gapic_third_party/BUILD.bazel:19:1: no such package '@io_grpc_protobuf//jar': Failed to fetch Maven dependency: Could not transfer artifact io.grpc:grpc-protobuf:jar:1.16.1 from/to io_grpc_protobuf (https://repo1.maven.org/maven2/): connect timed out and referenced by '@gapic_third_party//:grpc'
ERROR: Analysis of target '//:pkg' failed; build aborted: no such package '@io_grpc_protobuf//jar': Failed to fetch Maven dependency: Could not transfer artifact io.grpc:grpc-protobuf:jar:1.16.1 from/to io_grpc_protobuf (https://repo1.maven.org/maven2/): connect timed out
INFO: Elapsed time: 11.597s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (1 packages loaded, 2729 targets configured)
    Fetching @com_github_pkg_errors; Restarting. 11s
```

**Comment by pemgithub** [comment]
I think maybe the fix is to modify the GAPID bazel build files to use rules_jvm_external instead of maven_jar.  Though I'm not sure how...

Reference: https://stackoverflow.com/questions/56943540/bazel-tell-external-dependency-maven-to-use-local-jar

**Comment by AWoloszyn** [comment]
Sorry about the slow delay, I will take a look at this, but our build expert is away right now.

**Comment by null77** [comment]
Thanks, FYI I think my build would get farther by using the instructions above (choco install bazel --version=0.26.0) but I haven't tried it. I was reluctant to install bazel again because I have some custom Python and msysgit config in my environment that choco overwrote.

**Comment by AWoloszyn** [comment]
I am testing the build now. So far so good.
w.r.t. the maven dependencies. We will have to wait until @pmuetschard is back, he would know best, but you could stop building the UI for now, and see if that works for you.

```

copy_to(
    name = "pkg-lib",
    srcs = [
        "//core/vulkan/vk_virtual_swapchain/cc:json",
        "//core/vulkan/vk_virtual_swapchain/cc:libVkLayer_VirtualSwapchain",
#        "//gapic:gapic_deploy.jar",
        "//gapii/cc:libgapii",
        "//gapii/vulkan/vk_graphics_spy/cc:json",
    ],
 #   rename = {
 #       "gapic_deploy.jar": "gapic.jar",
  #  },
    to = "pkg/lib",
)

```

You can then copy gapic.jar from a release build of GAPID into `./bazel-bin/lib/gapic.jar`. I think that should be enough to finish the build without maven issues.

It is possible to also download the .jar files and point directly to them but that is a bit frustrating.

**Comment by AWoloszyn** [comment]
I just built this on windows using the instructions provided.

There was one step that I did that was not in the instructions, so it would be good to see if this fixes for anyone else.

I used bazel 0.26.0
Android NDK 0.18b 
`PATH=C:\tools\msys64\usr\bin;C:\tools\msys64\mingw64\bin;%PATH%` << This was slightly different than the instructions.
```
ANDROID_HOME=<path/to/android/sdk>
ANDROID_NDK_HOME=<path/to/android/ndk>
JAVA_HOME=<path_to_jdk8>
```

**Comment by pemgithub** [comment]
@AWoloszyn I commented out those linse in gapid\BUILD.bazel.  However, I still get the error:
`ERROR: C:/users/pemgithub/_bazel_pemgithub/3mdmxsrc/external/gapic_third_party/BUILD.bazel:41:1: no such package '@org_lwjgl_core_natives_windows//jar': Failed to fetch Maven dependency: Could not transfer artifact org.lwjgl:lwjgl:jar:natives-windows:3.2.0 from/to org_lwjgl_core_natives_windows (https://repo1.maven.org/maven2/): connect timed out and referenced by '@gapic_third_party//:lwjgl'
ERROR: Analysis of target '//:pkg' failed; build aborted: no such package '@org_lwjgl_core_natives_windows//jar': Failed to fetch Maven dependency: Could not transfer artifact org.lwjgl:lwjgl:jar:natives-windows:3.2.0 from/to org_lwjgl_core_natives_windows (https://repo1.maven.org/maven2/): connect timed out`

I'm able to "download the .jar files" using chrome.  For ex, io.grpc:grpc-protobuf:jar:1.16.1 maps to https://repo1.maven.org/maven2/io/grpc/grpc-protobuf/1.16.1/grpc-protobuf-1.16.1.jar.

However, I don't know how to "point directly to them"...  How do I tell bazel (or GAPID's bazel files) to point to my local copy of grpc-protobuf-1.16.1.jar etc?

**Comment by johnplate** [comment]
Fixed with #3243, requires Bazel 1.2.0.
